docs/proposals/diamond-design.md

```diff
@@ -0,0 +1,562 @@
+client _is_ an incremental upload. The other difference is how the user wants to deal with conflicts.
+
+Therefore, we need to capture user intent. We add some additional CLI flags to specify how datamon should deal
+with conflicts. See also [§ Handling Conflicts below](#handling-conflicts).
+
+
+Available options at commit time:
+```
+datamon diamond commit --diamond {diamond-id} \
+                      [--with-conflicts]   # <- this is the default - overlaps are reported as conflicts detected when merging splits
+
+datamon diamond commit --diamond {diamond-id} \
+                       --with-checkpoints   # <- indicates an incremental upload: conflicts are subsequent versions of uploaded files
+
+datamon diamond commit --diamond {diamond-id} \
+                       --no-conflicts       # <- forbids any overlap to take place at commit time
+
+datamon diamond commit --diamond {diamond-id} \
+                       --ignore-conflicts   # <- do not report about any conflict: intermediate versions
+```
+
+Available option at upload time:
+```
+datamon diamond add --diamond {diamond-id} \
+                    --split {split-id} --path {source}   #  <- locally ignores conflicts with previous instance of
+                                                         #     {split-id}, no change in behavior at commit time
+```
+
+Use `--no-conflicts` to guard incremental uploads against clobbering already uploaded files, whenever this is not desirable.
+Flags determining the conflict-handling mode cannot be used jointly.
+
+### Handling concurrency
+
+Premises:
+
+1. Several diamonds can run concurrently on the same repo, resulting in distinct bundles
+2. Within a single diamond, several `add` commands may run concurrently to contribute to a single same bundle
+3. Only one eventual commit operation may be carried out. Any attempt to compete on committing a diamond must be rejected.
+4. Some `split add` commands may fail, some may complete, some may be subject to a retry
+
+
+Design:
+
+1. Concurrent diamonds.
+  * A split may start **if and only if** the diamond metadata file is already existing in the vmetadata bucket.
+  * The uniqueness of the diamond ID allows for several diamonds to run in parallel.
+    Note that uniqueness is required only for any single repo.
+
+2. Concurrent splits.
+  * Each `split add` command runs with its uniquely generated internal ID: metadata can be safely modified within this namespace.
+
+3. Commit consistency.
+  * The diamond is in effect concluded with the first commit associated with it. No further action is allowed on a committed diamond.
+  * Any attempt to commit while another commit is in progress is rejected (the operation should normally be fast
+    since only metadata is affected).
+  * For the duration of the commit operation, the diamond is in effect locked in "committing" status.
+  * To recover from uncontrolled failures during a commit (e.g. pod crash,...) this lock is considered stale after some
```

Comment:
This is not needed, we only need a single done.yaml file in the path with the bundle ID in it to mark commit. The put if not present primitive of cloud storage APIs will provide the necessary protection.

Reply:
All that part about locking is indeed no more needed since the file no more needs to be locked.